File: packages/web/src/utils/encrypt.ts
/**
 * It takes an array of characters and a shift value and returns a map of characters to characters
 * @param {any} alphabets - an array of alphabets
 * @param {number} shift - the amount of shift to be applied to the alphabets
 * @returns A function that takes in a string and a shift number and returns a string.
 */
const createMAp = (alphabets: any, shift: number) => {
    return alphabets.reduce(
        (charsMap: any, currentChar: any, charIndex: any) => {
            const copy = { ...charsMap }
            let ind = (charIndex + shift) % alphabets.length
            if (ind < 0) {
                ind += alphabets.length
            }
            copy[currentChar] = alphabets[ind]
            return copy
        },
        {}
    )
}

/**
 * It takes a string and a shift value, and returns a new string with each character shifted by the
 * shift value
 * @param {any} org - The original string that you want to encrypt.
 * @param [shift=0] - The number of characters to shift the alphabet by.
 * @returns A function that takes in a string and a shift value and returns a string.
 */
const encrypt = (org: any, shift = 0) => {
    const alphabets =
        'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890'.split(
            ''
        )
    const map = createMAp(alphabets, shift)
    return org
        .toLowerCase()
        .split('')
        .map((char: any) => map[char] || char)
        .join('')
}

export default encrypt

File: packages/mobile/lib/classes/utils/encrypt.dart
/// We're creating a map from the list of alphabets, where the key is the current character and the
/// value is the character at the index of the current character plus the shift
///
/// Args:
///   alphabets (List<String>): A list of alphabets.
///   shift (int): The number of characters to shift the alphabet by.
///
/// Returns:
///   A map with the key being the current character and the value being the shifted character.
Map<String, String> createMap(List<String> alphabets, int shift) {
  return alphabets.asMap().map((charIndex, currentChar) {
    int ind = (charIndex + shift) % alphabets.length;
    if (ind < 0) {
      ind += alphabets.length;
    }
    return MapEntry(currentChar, alphabets[ind]);
  });
}

/// It encrypts a string using the Caesar cipher.
///
/// Args:
///   org (String): The string to be encrypted.
///   shift (int): The amount of shift to be applied to the alphabets. Defaults to 0
///
/// Returns:
///   A map of the alphabets with the shift applied.
String encrypt(String org, [int shift = 0]) {
  final alphabets =
      'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890'
          .split('');
  final map = createMap(alphabets, shift);
  return org.toLowerCase().split('').map((char) => map[char] ?? char).join('');
}
